driver/ni/ni.cpp

```diff
@@ -92,178 +514,357 @@ void ni::Source::getIndexKeys(){
 ni::Source::Source(
     TaskHandle task_handle,
     const std::shared_ptr<task::Context> &ctx,
-    const synnax::Task task): task_handle(task_handle), ctx(ctx), task(task){
+    const synnax::Task task) : task_handle(task_handle), ctx(ctx), task(task) {
 }
 
-void ni::Source::parseConfig(config::Parser &parser){
-    
-    LOG(INFO) << "[NI Source] Parsing Configuration for task " << this->reader_config.task_name;
+void ni::Source::parse_config(config::Parser &parser) {
     // Get Acquisition Rate and Stream Rates
-    this->reader_config.sample_rate = parser.required<uint64_t>("sample_rate");
-    this->reader_config.stream_rate = parser.required<uint64_t>("stream_rate");
+    this->reader_config.sample_rate.value = parser.required<uint64_t>("sample_rate");
+    this->reader_config.stream_rate.value = parser.required<uint64_t>("stream_rate");
     this->reader_config.device_key = parser.required<std::string>("device");
-    this->reader_config.timing_source = "none"; // parser.required<std::string>("timing_source"); TODO: uncomment this when ui provides timing source
-
-    LOG(INFO) << "[NI Source] retrieving device " << this->reader_config.task_name;
-
-    auto [dev, err] = this->ctx->client->hardware.retrieveDevice(this->reader_config.device_key);
-
-    LOG(INFO) << "[NI Source] device retrieved " << this->reader_config.task_name;
-
-    if (err != freighter::NIL) {
-        LOG(ERROR) << "[NI Reader] failed to retrieve device " << this->reader_config.device_name;
-        this->ok_state = false;
-        return;
+    this->reader_config.timing_source = "none";
+    // parser.required<std::string>("timing_source"); TODO: uncomment this when ui provides timing source
+    if (parser.optional<bool>("test", false))
+        this->reader_config.device_name = parser.required<std::string>(
+            "device_location");
+    else {
+        auto [dev, err] = this->ctx->client->hardware.retrieveDevice(
+            this->reader_config.device_key);
+        if (err != freighter::NIL) {
```

Comment:
The boolean operator on `freighter:Error` already compares it against nil. Just use `if(err)` instead